File: lab3/src/main/java/org/example/Class/Device.java
package org.example.Class;

import jakarta.persistence.*;
import lombok.*;
import com.fasterxml.jackson.annotation.JsonBackReference;
import java.io.Serializable;
import java.time.LocalDateTime;
import java.util.Map;
import java.util.HashMap;

@Entity
@Table(name = "device")
@Inheritance(strategy = InheritanceType.JOINED)
@Data
@NoArgsConstructor
@AllArgsConstructor
@EqualsAndHashCode(onlyExplicitlyIncluded = true)
public abstract class Device implements Serializable {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @EqualsAndHashCode.Include
    private Long id;

    @Column(name = "name", nullable = false, length = 100)
    private String name;

    @Column(name = "manufacturer", length = 100)
    private String manufacturer;

    @Column(name = "is_on")
    private boolean isOn;

    @Column(name = "created_at")
    private LocalDateTime createdAt;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "home_id")
    @JsonBackReference
    @ToString.Exclude
    private SmartHome smartHome;

    public Device(String name, SmartHome smartHome) {
        this.name = name;
        this.manufacturer = "Unknown";
        this.isOn = false;
        this.createdAt = LocalDateTime.now();
        this.smartHome = smartHome;
    }

    public Device(String name, String manufacturer, boolean isOn, SmartHome smartHome) {
        this.name = name;
        this.manufacturer = manufacturer;
        this.isOn = isOn;
        this.createdAt = LocalDateTime.now();
        this.smartHome = smartHome;
    }

    public void turnOn() {
        this.isOn = true;
        System.out.println("Устройство " + name + " включено");
    }

    public void turnOff() {
        this.isOn = false;
        System.out.println("Устройство " + name + " выключено");
    }

    public Map<String, Object> displayInfo() {
        Map<String, Object> data = new HashMap<>();
        data.put("type", "Устройство");
        data.put("id", id);
        data.put("name", name);
        data.put("manufacturer", manufacturer);
        data.put("isOn", isOn);
        data.put("status", isOn ? "Включено" : "Выключено");
        data.put("createdAt", createdAt != null ? createdAt.toString() : "-");
        return data;
    }

    public double calculatePowerConsumption() {
        return 0;
    }

    public abstract String getDeviceType();

    public abstract String getShortDescription();

    public abstract Map<String, Object> getDetailedInfo();
}

File: oopLaba2/Device.cs
using System;

namespace oopLaba2
{
    /// <summary>
    /// Базовый абстрактный класс для всех устройств умного дома
    /// </summary>
    public abstract class Device
    {
        // Собственные поля базового класса
        private string _name;
        private string _manufacturer;
        private bool _isOn;
        private DateTime _createdAt;

        // Конструктор по умолчанию
        protected Device()
        {
            _name = "fdsafawe";
            _manufacturer = "fdaswefda";
            _isOn = false;
            _createdAt = DateTime.Now;
        }

        // Конструктор с параметрами
        protected Device(string name, string manufacturer)
        {
            _name = name;
            _manufacturer = manufacturer;
            _isOn = false;
            _createdAt = DateTime.Now;
        }

        // Конструктор копирования
        protected Device(Device other)
        {
            _name = other._name;
            _manufacturer = other._manufacturer;
            _isOn = other._isOn;
            _createdAt = other._createdAt;
        }

        // Геттеры/Сеттеры для доступа к полям
        public string Name
        {
            get { return _name; }
            set { _name = value; }
        }

        public string Manufacturer
        {
            get { return _manufacturer; }
            set { _manufacturer = value; }
        }

        public bool IsOn
        {
            get { return _isOn; }
            set { _isOn = value; }
        }

        public DateTime CreatedAt
        {
            get { return _createdAt; }
            set { _createdAt = value; }
        }

        // Собственные методы базового класса
        public virtual void TurnOn()
        {
            _isOn = true;
            Console.WriteLine($"Устройство {_name} включено");
        }

        public virtual void TurnOff()
        {
            _isOn = false;
            Console.WriteLine($"Устройство {_name} выключено");
        }

        public virtual string GetStatus()
        {
            return $"Устройство: {_name}, Производитель: {_manufacturer}, Состояние: {(_isOn ? "Включено" : "Выключено")}, Создано: {_createdAt}";
        }

        // Перекрытые методы - виртуальные для полиморфизма
        public virtual string GetDeviceType()
        {
            return "Device";
        }

        public virtual string GetDetailedInfo()
        {
            return GetStatus();
        }

        // Абстрактный метод для обязательной реализации в дочерних классах
        public abstract string GetShortDescription();
    }
}